### ISSUE TITLE
`GPUBuffer`s shouldn't have to accumulate a list of `mapAsync` promises to reject

### ISSUE THREAD

**Issue by jimblandy** [gpuweb/gpuweb#3271]
If content makes a series of `mapAsync` requests without waiting for the resulting promises to be settled, since validation happens on the device timeline, the device will accept the first valid request, and reject all the others before and after it. Since the content timeline can't tell which of the promises is going to be accepted, it must retain all of them until it gets news from the device timeline. This seems unnecessary and unhelpful.

In a bit more detail:

Suppose content makes a series of `mapAsync` requests on a `GPUBuffer` in the "unmapped" state, without waiting for any of the promises to resolve, and without any intervening `unmap` requests. The language around `mapAsync` is under revision, but I believe even in the revised language, the first such `mapAsync` that passes validation puts the buffer in the "mapping pending" state, and all preceding and succeeding map requests get rejected.

I would like to suggest that the content timeline return immediately rejected promises for all `mapAsync` requests but the first, until that first request's promise is settled (either rejected or fulfilled). This would allow the `GPUBuffer` implementation to retain only a single promise, simplifying its state. This also seems to me to be a more predictable API: there is no possible use for a series of `mapAsync` calls without waiting on the promises.

This would remove the ability to fire off a series of `mapAsync` calls to see which one succeeds. That seems like a niche use case; the conditions under which `mapAsync` succeeds are pretty predictable.

**Comment by Kangz** [comment]
There's several reasons we mapping works like this today, and we may choose to revisit the decision (although personally I'd rather not as we're trying to aim for a V1, and that the current version is implementable somewhat easily).

 - Right now a single range of a buffer can be mapped but there were discussions in the past about allowing multiple different ranges to be mapped at the same time. If we do the change to make validation that the buffer is unmapped on the content side, how would this work?
 - In the future the buffers being mappable or not could depend on more data. For example whether there is a UMA extension, what flags are passed to `mapAsync`, what limits are set (for potential alignment etc). Right now knowing whether a mapAsync will succeed is easy, but it might not hold in the future. Only the device timeline has the full picture.
 - The buffer mapping is a special case, but all other validation is deferred as much as possible to the device timeline. Adding more client-side state that's not roundtripped to the device timeline would make buffer mapping even more special and have even more code duplication between the client and server side of WebGPU.

**Comment by kainino0x** [comment]
There is quite a bit of complexity here due to client-server model, but most of the problems here are caused by client-server state not being clearly defined. In the spec as written it's only possible to have one pending `Promise` stored on the GPUBuffer object as `[[mapping]]`; this is the one that is cancelled by `unmap()`. Other promises that can be rejected by failing mapAsync validation aren't held onto directly, they're only held by "closure" from the mapAsync callback.

Fundamentally, I don't think there's anything wrong with holding things by closure. In Chromium, we can do this pretty literally, with a callback object that holds a reference to the Promise object.

I think the *current draft* of #3014 actually would [resolve](https://github.com/gpuweb/gpuweb/pull/3014/commits/d6630c3619af33ce691188b55c8acaf899caed74) this issue. In this PR, the client vs server state is partly fixed up: mapping states are now tracked separately on both the client and server (where previously it wasn't clear what was on the client vs server). As written, `[[mapping]]` is clearly a client state and `mapAsync` actually checks the `mapState` since it needs to know whether it should set the `[[mapping]]` slot or not.

Technically, I don't think we actually need to hold onto any of the promises explicitly, and this alternative might make more sense than what I have in #3014. `[[mapping]]` is held onto for `unmap()` because the S->C map message and C->S unmap message can cross each other, resulting in unmap() arriving at the server after it has already sent the "map success" message. However just knowing the `mapState` on the client is probably enough: If the client receives "map success" while the state is "unmapped" then it knows to ignore it. If the server goes to send a "map success" message but sees that the map has been cancelled with `unmap()`, it can send a "map cancelled" message instead.

For this reason I think I'm going to remove this from #3014 for now and leave an issue pointing back to this issue (thanks for filing and prompting me to think about this more).

Note #3251 which also deals with mapping state and `unmap()`.

**Comment by kainino0x** [comment]
I think it's even a bit simpler than I described. Here's a diagram showing early-reject vs late-reject for cancellation.

https://docs.google.com/drawings/d/1tskzoB11nJm11iD7-uSogglTW32QccStokOLxDMDBoQ/edit?usp=sharing

**Comment by kainino0x** [comment]
In conclusion:
- I agree with Corentin that trying to do all the mapAsync validation on the client is impractical. In the best case it'll mean we have to mirror a number of states on the "client" side, in a thread-shared way. More fundamentally, only the device timeline knows if a buffer was successfully created, and currently mapping is not allowed on invalid buffers. (Though I think mapping could be allowed on invalid buffers - https://github.com/gpuweb/gpuweb/issues/1629#issuecomment-1144405539)
- I do think we should move the rejection from unmap() to mapAsync-response, and stop holding onto the promise statefully on the client. This would be a change relative to the current spec.

**Comment by Kangz** [comment]
> I think it's even a bit simpler than I described. Here's a diagram showing early-reject vs late-reject for cancellation.
>
> https://docs.google.com/drawings/d/1tskzoB11nJm11iD7-uSogglTW32QccStokOLxDMDBoQ/edit?usp=sharing

I think there might still be a slight ordering problem if we just check the client-side state if we do the following sequence:

 - Client: mapAsync that's valid
 - Server: queue passes and we send back the mapping.
 - Client: unmap
 - Client: remap an invalid mapAsync
 - Client: receive the successful mapping, believe it is for the second mapping when instead it is for the first.

Am I missing something?

**Comment by kainino0x** [comment]
You're right. I think generation numbers are required, or something similar (e.g. comparing the promise reference from the closure with the promise reference held as state on the GPUBuffer client object). I think both alternatives have this problem though, so it doesn't really affect the difference between the two.

https://docs.google.com/drawings/d/1uunE5-IWSaCWN9gseq6Zls-M813SprwT5udB4iil9WA/edit?usp=sharing

**Comment by kainino0x** [comment]
I dug a bit into how Chromium handles this. As I said before, we don't hold onto promises for the purpose of resolving them - those are held onto by "closure" in a callback. However, something does have to hold onto that closure. IIUC the Dawn Wire client holds onto all in-flight mapping requests. These get resolved or rejected based on responses from the server, and also get [early rejected](https://source.chromium.org/chromium/chromium/src/+/main:third_party/dawn/src/dawn/wire/client/Buffer.cpp;l=344;drc=04f12142bd5de7ed48555978012df4577305e4ce) in unmap(). This list can include ones that are destined to fail because of the server-side state check to check that the buffer isn't already mapped. Note Dawn (webgpu.h) guarantees all map callbacks are called exactly once so they can be used for memory management (you can free in the callback, and this won't cause leaks or double-frees).

Holding the full list as Dawn does, and as mentioned in the issue, seems fine. But I do think that, in the current one-map-at-a-time model, it's possible to reduce this list to just 0-1 callbacks with additional state checks on the client - essentially, early reject if there's already a callback in flight. However this wouldn't get us much other than just getting rid of the list. Because the service doesn't trust the client, the service has to handle mapAsync calls coming in at weird times anyway. Adding this client-side check would make the handling of that dead code with functioning clients, but live code for compromised clients. This is a bit scary to me since that code will be undertested, but has to exist so the server code doesn't get any simpler. The client code does get a bit simpler but I don't think it's a lot.

As for the spec, looking at our actual code helped me figure out that our current spec really should hold a list of promises instead of just one. At least, if we want to model the behavior that Chromium currently has.

**Comment by Kangz** [comment]
> These get resolved or rejected based on responses from the server, and also get [early rejected](https://source.chromium.org/chromium/chromium/src/+/main:third_party/dawn/src/dawn/wire/client/Buffer.cpp;l=344;drc=04f12142bd5de7ed48555978012df4577305e4ce) in unmap().

Note: that code only overrides the future status given by the server with "unmapped-before-success" so that even if there's a validation error or something else, we get the correct status from the point of view of the client. But the rest of the analysis makes sense to me.

**Comment by jimblandy** [comment]
> Fundamentally, I don't think there's anything wrong with holding things by closure. In Chromium, we can do this pretty literally, with a callback object that holds a reference to the Promise object.

I agree with this. In Firefox, as well, the promise that `mapAsync` creates gets captured by a closure owned by the IPC system, which is called when the buffer is ready. Having many closures retained that way seems fine to me.

But since promises are one-to-one with `mapAsync` calls, it seems to me that the state of the promise itself provides a clean way to handle this:

- Let `mapAsync` save the promise it creates on the buffer. If the buffer already has a saved promise from a prior call, `mapAsync` should immediately reject that promise, and replace it with the one created for this call.
- Let `unmap` reject the saved promise and clear the buffer's reference to it.
- Let the "mapped" notification closure check the state of its captured promise. If it has already been settled, ignore the notification. Otherwise, fulfill the promise, clear the buffer's reference to it, and mark the buffer mapped.

With these rules, there is at most one unsettled promise at a time: the one currently saved on the buffer.

Since `unmap` rejects the most recent `mapAsync` call's promise, the "mapped" notification corresponding to that call will always be ignored. It's obvious that we'll ignore exactly those notifications that we should.

Every `mapAsync` or `unmap` call can be unconditionally transmitted to the GPU process, so its validation paths can all be exercised.

In the content process, `mapAsync`, `unmap`, and "mapped" notifications affect the buffer's state in the natural way. In the GPU process, the buffer's state is determined by the last `map` or `unmap` message it received from content.

While there is no means for content JS to test whether a promise is settled or not, Firefox's C++ representation of the promise does have a method that does so, and I assume Chromium's promises do too.

**Comment by nical** [comment]
Firefox's buffer mapping logic is currently very similar to Dawn's with the exception that conceptually it only tracks up to one promise, and rejects other promises early on the content side if there is already one in flight. We also had a version of the code that tracked an array of promises. It's very similar but what jim proposes is a bit simpler and, I think, models in a straightforward way the expected use of the current state of WebGPU.

> Right now a single range of a buffer can be mapped but there were discussions in the past about allowing multiple different ranges to be mapped at the same time. If we do the change to make validation that the buffer is unmapped on the content side, how would this work?

I believe that Jim is operating under the principle that the spec and implementation should be as simple as possible and complexity should be added when needed instead of provisioning for potential future.
Currently the array is needed to handle a very niche pattern (firing multiple map requests until one succeds). If/when we support multiple map ranges, my understanding is that relaxing the proposed addendum (not rejecting the second promise right away) would be backwards compatible. Is that correct?

> In the future the buffers being mappable or not could depend on more data. For example whether there is a UMA extension, what flags are passed to mapAsync, what limits are set (for potential alignment etc). Right now knowing whether a mapAsync will succeed is easy, but it might not hold in the future. Only the device timeline has the full picture.

I believe that we all agree on this. We also want as much of the validation to be on the device side as possible/convenient. This proposal is about allowing the implementation to simplify the tracking of map/unmap requests for now and avoiding the need to track multiple requests until doing so provides value.

> The buffer mapping is a special case, but all other validation is deferred as much as possible to the device timeline. Adding more client-side state that's not roundtripped to the device timeline would make buffer mapping even more special and have even more code duplication between the client and server side of WebGPU.

I agree in principle. Jim's proposal aims at reducing the amount of client side state in the implementations. It is state that isn't explicitly worded by the spec but currently has to exist to support it which isn't much better.

I realize that the proposal is quite nitpicky. We wouldn't have bothered if it wasn't a very low hanging fruit.

**Comment by kainino0x** [comment]
I agree with everything in @jimblandy and @nical's comments and am totally fine with changing the spec in this way. No strong feelings.

(this is on the agenda for Wednesday)

**Comment by kainino0x** [comment]
Accepted
